vespalib/src/vespa/vespalib/process/process.cpp

```diff
@@ -0,0 +1,153 @@
+// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+
+#include "process.h"
+#include "pipe.h"
+#include "close_all_files.h"
+
+#include <vespa/vespalib/util/size_literals.h>
+#include <vespa/vespalib/util/require.h>
+
+#include <sys/types.h>
+#include <sys/wait.h>
+
+#include <unistd.h>
+#include <fcntl.h>
+
+namespace vespalib {
+
+Process::Process(const vespalib::string &cmd, bool capture_stderr)
+  : _pid(-1),
+    _in(),
+    _out(),
+    _in_buf(4_Ki),
+    _out_buf(4_Ki),
+    _eof(false)
+{
+    Pipe pipe_in = Pipe::create();
+    Pipe pipe_out = Pipe::create();
+    REQUIRE(pipe_in.valid() && pipe_out.valid());
+    pid_t pid = fork();
+    REQUIRE(pid != -1);
+    if (pid == 0) {
+        dup2(pipe_in.read_end.fd(), STDIN_FILENO);
+        dup2(pipe_out.write_end.fd(), STDOUT_FILENO);
+        if (capture_stderr) {
+            dup2(pipe_out.write_end.fd(), STDERR_FILENO);
+        } else {
+            int dev_null = open("/dev/null", O_WRONLY);
```

Comment:
Might need to check for failure. if dev_null < 0 then open failed, typical case when we have run out of available file descriptors.

Reply:
what should we do in that case? I figured dup2/close will be 'ignored' if open returns -1, which is 'best effort'

Reply:
I agree, both dup2() and close() will return EBADF.